### ISSUE TITLE
Implement a `strptime` method

### ISSUE THREAD

**Issue by spencerkclark** [Unidata/cftime#277]
`cftime.datetime` objects implement a `strftime` method, which one can use to convert a datetime to a string with a given format.  It would be convenient if they also implemented the inverse method, [`strptime`](https://docs.python.org/3/library/datetime.html#datetime.datetime.strptime), which allows one to parse a string with a specified format into a datetime object.  I would imagine something like this:

```
>>> cftime.datetime.strptime("2000-01-01", "%Y-%m-%d", calendar="julian")
cftime.datetime(2000, 1, 1, 0, 0, 0, 0, calendar='julian', has_year_zero=False)
```

It is kind of a handful, but [the implementation in cpython](https://github.com/python/cpython/blob/d44815cabc0a8d9932df2fa95cb374eadddb7c17/Lib/_strptime.py#L309-L557) might serve as a useful reference.  @jswhit I would be happy to help contribute this if you agree it would be a useful addition to cftime.

cc: @nbren12

**Comment by nbren12** [comment]
+1 on this? Is there some reason you couldn't just use `datetime.datetime.strptime` under the hood?

**Comment by spencerkclark** [comment]
> Is there some reason you couldn't just use datetime.datetime.strptime under the hood?

Different calendars.  Some cftime dates are not valid in the proleptic Gregorian calendar (e.g. February 30th in a 360 day calendar).

**Comment by nbren12** [comment]
gotcha! forgive my ignorance on these things.

**Comment by nbren12** [comment]
The cpython code is a bit long but not so bad. Defines a single regexp and then harvests the matched groups from it converting to int as necessary. There is some weird thread-safe caching logic to make sure re.compile isn't too slow, but that doesn't seem worth replicating.

**Comment by spencerkclark** [comment]
All good, yeah, I think it would largely be copy-pastable.  I agree regarding the caching logic.

**Comment by jswhit** [comment]
@spencerkclark I think this would be a great feature to have

**Comment by spencerkclark** [comment]
Awesome, thanks @jswhit for confirming.  Indeed I saw your thumbs up earlier, but have just been caught up with other things lately.  Hopefully I'll have a chance to return to this soon.

**Comment by jswhit** [comment]
or we could use `datetime.datetime.strptime` and then convert the result from the `proleptic_gregorian` calendar using `cftime.datetime.change_calendar`.

**Comment by jswhit** [comment]
actually, `change_calendar` is not even needed.  Any reason something like this wouldn't work?

```python
def strptime(datestring, format, calendar='standard', has_year_zero=None):
    """
    Return a datetime corresponding to date_string, parsed according to format,
    with a specified calendar and year zero convention.
    """
    # use python's datetime.strptime to get a python datetime instance
    # (using proleptic_gregorian calendar)
    pydatetime = datetime_python.strptime(datestring, format)
    # account for utcoffset
    if getattr(pydatetime, 'tzinfo',None) is not None:
        pydatetime = pydatetime.replace(tzinfo=None) - pydatetime.utcoffset()
    # convert the cftime datetime instance
    return datetime(pydatetime.year, pydatetime.month, pydatetime.day,
                    pydatetime.hour, pydatetime.minute, pydatetime.second,
                    pydatetime.microsecond, calendar=calendar, has_year_zero=has_year_zero)
```

**Comment by spencerkclark** [comment]
The issue is that something like `datetime_python.strptime("2000-02-30", "%Y-%m-%d")` will fail, because it is not a valid date in the proleptic Gregorian calendar, but ideally we'd support it in the 360-day calendar.

**Comment by jswhit** [comment]
Yes, that is an issue.  The above approach will only work for 'real-world' calendars with dates that exist in the proleptic_gregorian calendar.  That probably covers the majority of use cases though.

**Comment by jswhit** [comment]
closed by PR #289
